File: implementations/swift/oshash.swift
// OpenSubtitles Hash (OSHash) - Swift implementation.
// Algorithm: file_size + sum_le_u64(first_64KB) + sum_le_u64(last_64KB)
import Foundation

let chunkSize = 65536

func computeHash(_ path: String) -> String {
    let url = URL(fileURLWithPath: path)
    let fileHandle = try! FileHandle(forReadingFrom: url)
    defer { fileHandle.closeFile() }

    fileHandle.seekToEndOfFile()
    let fileSize = fileHandle.offsetInFile
    fileHandle.seek(toFileOffset: 0)

    var hash: UInt64 = fileSize

    // Hash first 64KB
    let headData = fileHandle.readData(ofLength: chunkSize)
    headData.withUnsafeBytes { raw in
        let longs = raw.bindMemory(to: UInt64.self)
        for val in longs { hash &+= UInt64(littleEndian: val) }
    }

    // Hash last 64KB
    fileHandle.seek(toFileOffset: max(0, fileSize - UInt64(chunkSize)))
    let tailData = fileHandle.readData(ofLength: chunkSize)
    tailData.withUnsafeBytes { raw in
        let longs = raw.bindMemory(to: UInt64.self)
        for val in longs { hash &+= UInt64(littleEndian: val) }
    }

    return String(format: "%016lx", hash)
}

guard CommandLine.arguments.count > 1 else {
    fputs("Usage: oshash <file>\n", stderr)
    exit(1)
}
print(computeHash(CommandLine.arguments[1]))

File: implementations/dart/oshash.dart
// OpenSubtitles Hash (OSHash) - Dart implementation.
import 'dart:io';
import 'dart:typed_data';

const chunkSize = 65536;

String computeHash(String filepath) {
  final file = File(filepath);
  final raf = file.openSync(mode: FileMode.read);
  final filesize = file.lengthSync();

  if (filesize < chunkSize * 2) {
    raf.closeSync();
    throw Exception('File too small: $filesize bytes');
  }

  // Use wrapping arithmetic via masking
  int hash = filesize;

  // Hash first 64KB
  final headBytes = Uint8List.fromList(raf.readSync(chunkSize));
  final headData = ByteData.sublistView(headBytes);
  for (var i = 0; i < chunkSize ~/ 8; i++) {
    // Read as signed int64 LE (Dart int is signed 64-bit)
    hash += headData.getInt64(i * 8, Endian.little);
  }

  // Hash last 64KB
  raf.setPositionSync(filesize - chunkSize);
  final tailBytes = Uint8List.fromList(raf.readSync(chunkSize));
  final tailData = ByteData.sublistView(tailBytes);
  for (var i = 0; i < chunkSize ~/ 8; i++) {
    hash += tailData.getInt64(i * 8, Endian.little);
  }

  raf.closeSync();

  // Convert to unsigned hex representation
  // Dart int wraps at 64 bits, so just mask and format
  int unsigned = hash;
  if (unsigned < 0) {
    // Convert negative signed to unsigned equivalent via BigInt
    BigInt big = BigInt.from(hash).toUnsigned(64);
    return big.toRadixString(16).padLeft(16, '0');
  }
  return unsigned.toRadixString(16).padLeft(16, '0');
}

void main(List<String> args) {
  if (args.isEmpty) {
    stderr.writeln('Usage: dart oshash.dart <file>');
    exit(1);
  }
  print(computeHash(args[0]));
}
